ibis/expr/types/relations.py

```diff
@@ -829,9 +829,12 @@ def difference(self, *tables: Table, distinct: bool = True) -> Table:
         >>> t1.difference().equals(t1)
         True
         """
-        return functools.reduce(
+        t = functools.reduce(
             functools.partial(ops.Difference, distinct=distinct), tables, self.op()
         ).to_expr()
+        if t.equals(self):
+            return t
+        return t.select(self.columns)
```

Comment:
So we effectively remove the table type from field selections?

Reply:
Not sure I follow what you're asking. The difference between this PR and upstream is that in this PR we always project the results of a set operation. The union is still underneath the projection.

Reply:
Ideally `t.select()` should be able to optimize this out in the future as a no-op. 

It feels like that we have more serious problems in the analysis code actually causing this bug.

Reply:
Agree, we should be able to handle this elsewhere, but for now this fixes the issue without too much hassle.

Reply:
Created #5917 for this.